lib/BuildSystem/BuildSystem.cpp

```diff
@@ -585,13 +588,34 @@ class DirectoryInputNodeTask : public Task {
                /*inputID=*/0);
   }
 
+  virtual void start(TaskInterface ti) override {
+    for (auto mustScanAfterPath: node.getMustScanAfterPaths().getValues()) {
+      ++totalBlockingDeps;
+      ti.request(BuildKey::makeNode(mustScanAfterPath).toData(), totalBlockingDeps);
+    }
+
+    if (totalBlockingDeps == 0) {
+      performUnblockedRequest(ti);
+    }
+  }
+
   virtual void providePriorValue(TaskInterface,
                                  const ValueType& value) override {
   }
 
-  virtual void provideValue(TaskInterface, uintptr_t inputID,
-                            const ValueType& value) override {
-    directorySignature = value;
+  virtual void provideValue(TaskInterface ti, uintptr_t inputID,
+                            const ValueType& valueData) override {
+    if (inputID == 0) {
+      directorySignature = valueData;
+    } else {
+      auto value = BuildValue::fromData(valueData);
+      // FIX ME (critical before merge): must check provided values are not erroneous
```

Comment:
In `DirectoryContentsTask::inputsAvailable` we have the following checks for corner cases:

```
    if (directoryValue.isMissingInput()) {
      ti.complete(BuildValue::makeMissingInput().toData());
      return;
    }

    if (directoryValue.isMissingOutput()) {
      // Rewrite a missing output as a missing input build value for this node, both conceptually and as a hedge against violating the downstream expectations of other rules.
      ti.complete(BuildValue::makeMissingInput().toData());
      return;
    }

    if (directoryValue.isFailedInput()) {
      ti.complete(BuildValue::makeFailedInput().toData());
      return;
    }

    // The input directory may be a 'mkdir' command, which can be cancelled or
    // skipped by the engine or the delegate. rdar://problem/50380532
    if (directoryValue.isSkippedCommand()) {
      ti.complete(BuildValue::makeSkippedCommand().toData());
      return;
    }
```

Reply:
- [ ] It'd be nice to have a test for erroneous cases too (e.g. "what happens when subpath of a directory input node is produced by a command that exits in error?")